nixos/lib/test-driver/test-driver.py

```diff
@@ -525,6 +526,36 @@ def screenshot(self, filename: str) -> None:
             if ret.returncode != 0:
                 raise Exception("Cannot convert screenshot")
 
+    def copy_from_vm(self, source: str, target_dir: str = "") -> None:
+        """Copy a file from the VM (specified by an in-VM source path) to $out
+        (with an optional subdirectory parameter). The file is copied via the
+        shared directory.
+        """
+        # Compute the source, target, and intermediate shared file names
+        out_dir = os.environ.get("out", os.getcwd())
+        filename = re.sub(r".*/", "", source)
```

Comment:
Please use `pathlib` here to calculate basenames and assemble paths:

https://docs.python.org/3/library/pathlib.html#methods-and-properties